pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,412 @@
+			c.attachCertToSecret(cert, syncCtx.Recorder())
+			// we've got the cert, stop progressing
+			return pbool(false), nil
+		}
+
+		if !shouldResubmit { // keep on waiting for the current CSR
+			return pbool(true), nil
+		}
+
+	}
+
+	// either we did not have a key or the CSR failed
+	if shouldResubmit {
+		if err := c.resubmitCSR(ctx, key); err != nil {
+			return nil, err
+		}
+		return pbool(true), nil
+	}
+
+	return pbool(false), nil
+}
+
+func (c *webhookAuthenticatorCertRequester) checkKeyCertSecret(ctx context.Context, recorder events.Recorder) (key *rsa.PrivateKey, secretComplete, shouldResubmit bool, err error) {
+	key, cert, err := c.getKeyCertFromSecret()
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, false, false, err
+	}
+
+	if key != nil && cert != nil {
+		// check that the certificate is not expiring
+		certValidityDuration := cert.NotAfter.Sub(cert.NotBefore)
+		if time.Now().Before(cert.NotBefore.Add(time.Duration(0.8 * float64(certValidityDuration)))) {
+			return key, true, false, nil
+		}
+		// set key == nil to trigger new key generation
+		key = nil
+
+		expiryMsg := fmt.Sprintf("the authenticator's certificate is almost expired (%s), will schedule a new CSR creation", cert.NotAfter.String())
+		recorder.Eventf("ClientCertExpiringSoon", expiryMsg)
+		klog.V(5).Infof(expiryMsg)
+	}
+
+	if key != nil {
+		return key, false, false, err
+	}
+
+	// either the key has always been nil, or the certificate is expiring
+	key, err = createSecretWithNewKey(c.secretClient, recorder)
+	return key, false, true, err
+
+}
+
+func (c *webhookAuthenticatorCertRequester) handleCSR(key *rsa.PrivateKey, recorder events.Recorder) (cert []byte, shouldResubmit bool, err error) {
+	csrObj, err := c.csrLister.Get(csrName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, true, nil
+		}
+		return nil, false, err
+	}
+
+	if len(csrObj.Status.Conditions) == 0 {
+		return nil, false, nil
+	}
+
+	for _, cond := range csrObj.Status.Conditions {
+		switch cond.Type {
+		case certapiv1.CertificateDenied, certapiv1.CertificateFailed:
+			if cond.Status == corev1.ConditionTrue {
+				recorder.Eventf("CSRFailure", "previous CSR got denied or it otherwise failed: {%s - %s: %s}", cond.Type, cond.Reason, cond.Message)
+				klog.Warning("the previous CSR got denied or it otherwise failed: %v", cond)
+				return nil, true, nil
+			}
+		}
+	}
+
+	for _, cond := range csrObj.Status.Conditions {
+		if cond.Type == certapiv1.CertificateApproved && cond.Status == corev1.ConditionTrue {
+			csrCert := csrObj.Status.Certificate
+			if len(csrCert) == 0 {
+				return nil, false, nil
+			}
+
+			certObj, err := pemDecodeCert(csrCert)
+			if err != nil {
+				return nil, false, fmt.Errorf("failed to parse the certiticate retrieved from CSR: %w", err)
+			}
+
+			if !key.PublicKey.Equal(certObj.PublicKey) {
+				recorder.Eventf("CSRBadCertKey", "the public key of the CSR certificate did not match the current private key")
```

Comment:
events based on state are bad

Reply:
if I remove it, we're likely to never find out why the rotation was triggered :|